src/network/operator.rs

```diff
@@ -0,0 +1,335 @@
+        }
+
+        if E::NODE_TYPE == NodeType::Operator {
+            if let Some(recipient) = operator.address {
+                // Initialize an update loop for the block template.
+                let operator = operator.clone();
+                let (router, handler) = oneshot::channel();
+                tasks.append(task::spawn(async move {
+                    // Notify the outer function that the task is ready.
+                    let _ = router.send(());
+                    // TODO (julesdesmit): Add logic to the loop to retarget share difficulty.
+                    loop {
+                        // Determine if the current block template is stale.
+                        let is_template_stale = match &*operator.block_template.read().await {
+                            Some(template) => operator.ledger_reader.latest_block_height().saturating_add(1) != template.block_height(),
+                            None => true,
+                        };
+
+                        // Update the block template if it is stale.
+                        if is_template_stale {
+                            // Construct a new block template.
+                            let transactions = operator.memory_pool.read().await.transactions();
+                            let (block_template, _) = operator
+                                .ledger_reader
+                                .get_block_template(recipient, E::COINBASE_IS_PUBLIC, &transactions, &mut thread_rng())
+                                .expect("Should be able to generate a block template");
+
+                            // Acquire the write lock to update the block template.
+                            *operator.block_template.write().await = Some(block_template);
+                        }
+
+                        // Sleep for `5` seconds.
+                        tokio::time::sleep(Duration::from_secs(5)).await;
+                    }
+                }));
+                // Wait until the operator handler is ready.
+                let _ = handler.await;
+            } else {
+                error!("Missing operator address. Please specify an Aleo address in order to operate a pool");
+            }
+        }
+
+        Ok(operator)
+    }
+
+    /// Returns an instance of the operator router.
+    pub fn router(&self) -> OperatorRouter<N> {
+        self.operator_router.clone()
+    }
+
+    /// Returns all the shares in storage.
+    pub fn to_shares(&self) -> Vec<(u32, HashMap<Address<N>, u64>)> {
+        self.state.to_shares()
+    }
+
+    ///
+    /// Performs the given `request` to the operator.
+    /// All requests must go through this `update`, so that a unified view is preserved.
+    ///
+    pub(super) async fn update(&self, request: OperatorRequest<N>) {
+        match request {
+            OperatorRequest::PoolRegister(peer_ip, address) => {
+                if let Some(block_template) = &*self.block_template.read().await {
```

Comment:
```suggestion
                if let Some(block_template) = self.block_template.read().await.cloned() {
```
in order not to hold the `read` for the duration of the whole block underneath